### PR TITLE
Adds an example about FunctionTransformer

### PR DESCRIPTION
**Pull request by xadupre**
Answer to issue #609.

### CODE REVIEW

File: docs/tutorial/plot_function_transformer.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        pass
+
+    def calculate_overprice(self, x, y):
+        return 100 * (x - y) / y
+
+    def fit(self, X, y=None):
+        return self
+
+    def transform(self, X, y=None):
+        x = X.apply(lambda x: self.calculate_overprice(x.a, x.b), axis=1)
```
Comment (paranjapeved15) [inline docs/tutorial/plot_function_transformer.py:85]: I would like to return all other columns (a,b,f) in the data frame along with the new column "c"

File: docs/tutorial/plot_function_transformer.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+# The first function returns the expected type and shape.
+
+
+def overprice_shape_calculator(operator):
+    check_input_and_output_numbers(operator, input_count_range=1, output_count_range=1)
+    # Gets the input type, the transformer works on any numerical type.
+    input_type = operator.inputs[0].type.__class__
+    # The first dimension is usually dynamic (batch dimension).
+    input_dim = operator.inputs[0].get_first_dimension()
+    operator.outputs[0].type = input_type([input_dim, 1])
```
Comment (paranjapeved15) [inline docs/tutorial/plot_function_transformer.py:127]: to return all 4 columns, something needs to change in the operator outputs shape, correct?

File: docs/tutorial/plot_function_transformer.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    # 100 * (x-y)/y  --> 100 * (X[0] - X[1]) / X[1]
+
+    zero = np.array([0], dtype=np.int64)
+    one = np.array([1], dtype=np.int64)
+    two = np.array([2], dtype=np.int64)
+    hundred = np.array([100], dtype=np.float32)
+
+    # Slice(data, starts, ends, axes)
+    x0 = OnnxSlice(X, zero, one, one, op_version=opv)
```
Comment (paranjapeved15) [inline docs/tutorial/plot_function_transformer.py:144]: I feel like the learning curve to convert normal operations to onnx operations is pretty steep. Could you suggest any good tutorials / docs to get more familiar with this?

File: docs/tutorial/plot_function_transformer.py (hunk 1 above)
Comment (xadupre) [reply docs/tutorial/plot_function_transformer.py:85]: I updated the pipeline to retain those two: `("ab", FunctionTransformer(), ["a", "b"]),`

File: docs/tutorial/plot_function_transformer.py (hunk 2 above)
Comment (xadupre) [reply docs/tutorial/plot_function_transformer.py:132]: No, this transformer takes 2 inputs and produces one. I added the others in other piece in the pipeline.

File: docs/tutorial/plot_function_transformer.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+def overprice_shape_calculator(operator):
+    check_input_and_output_numbers(operator, input_count_range=1, output_count_range=1)
+    # Gets the input type, the transformer works on any numerical type.
+    input_type = operator.inputs[0].type.__class__
+    # The first dimension is usually dynamic (batch dimension).
+    input_dim = operator.inputs[0].get_first_dimension()
+    operator.outputs[0].type = input_type([input_dim, 1])
+
+
+def overprice_converter(scope, operator, container):
```
Comment (paranjapeved15) [inline docs/tutorial/plot_function_transformer.py:135]: could you also please replace all occurrences of the word "overprice" with something else? I forgot to rename it while posting

File: docs/tutorial/plot_function_transformer.py (hunk 3 above)
Comment (xadupre) [reply docs/tutorial/plot_function_transformer.py:149]: For this example, I chose the API implemented in sklearn-onnx but I could extend the example to make it work with other API you prefer (see [Many ways to implement a custom graph in ONNX](https://sdpython.github.io/doc/onnx-array-api/dev/tutorial/onnx_api.html)) to write onnx graphs.

That said, I understand it is not easy. It takes time to rewrite an existing code with this list of [ONNX Operators](https://onnx.ai/onnx/operators/).

The easiest way is probably to use pytorch but it means to take a dependency on pytorch. The numpy API for onnx is less difficult but I initially started it for small functions. All the other options require to rewrite a function with onnx operators. Most of the operators follow numpy api. A couple of important points: there is no inplace modification allowed in onnx, notation [..., :, ...] can usually be handled with operator Slice but not always, Loop and Tests should be avoided.

I wrote this tutorial about onnx: https://onnx.ai/onnx/intro/python.html.

Comment (paranjapeved15) [reply docs/tutorial/plot_function_transformer.py:149]: Why is using PyTorch easier than sklearn for converting to onnx?

Comment (xadupre) [reply docs/tutorial/plot_function_transformer.py:155]: It is not necessary to rewrite your custom code with pytorch. With sklearn, the numpy code must be rewritten. I added a version with onnxscript as well.

File: docs/tutorial_2_new_converter.rst [hunk 5]
```diff
@@ -42,6 +42,7 @@ an example.
 
     auto_tutorial/plot_icustom_converter
     auto_tutorial/plot_jcustom_syntax
+    auto_tutorial/plot_function_transformer
```
Comment (fatcat-z) [inline docs/tutorial_2_new_converter.rst:45]: Nit:
Move it up so the names are alphabetically.

Comment (xadupre) [reply docs/tutorial_2_new_converter.rst:45]: I renamed it to keep that order. It makes more sense to keep it at this position.

**Review by paranjapeved15 — APPROVED** [review]

File: docs/tutorial/plot_jfunction_transformer.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+def growth_shape_calculator(operator):
+    check_input_and_output_numbers(operator, input_count_range=1, output_count_range=1)
+    # Gets the input type, the transformer works on any numerical type.
+    input_type = operator.inputs[0].type.__class__
+    # The first dimension is usually dynamic (batch dimension).
+    input_dim = operator.inputs[0].get_first_dimension()
+    operator.outputs[0].type = input_type([input_dim, 1])
+
+
+def growth_converter(scope, operator, container):
```
[the remark this answers is not in the document]
Comment (paranjapeved15) [reply docs/tutorial/plot_jfunction_transformer.py:141]: Thanks @xadupre . I have the following transformer: 
class Clipper(BaseEstimator, TransformerMixin):
    def __init__(self,lower_limit,upper_limit):
        self.lower_limit = lower_limit
        self.upper_limit = upper_limit
        pass

    def clip(self,data):
        return np.clip(data,self.lower_limit,self.upper_limit)

    def fit(self, X, y=None):
        return self

    def transform(self,X,y=None):
        x = X.apply(lambda x: self.clip(x), axis=1)
        return x.values.reshape((-1, 1))

    def get_feature_names_out(self):
            pass
Questions:
1. Is there a way to write a converter which can accept the lower limit and upper limit arguments (similar to my init function)? 
2. In your example below, you use operator.inputs[0]. What other contents does the inputs list contain?
